packages/autoloader/src/AutoloadGenerator.php

```diff
@@ -193,17 +198,65 @@ protected function parseAutoloadsType( array $packageMap, $type, PackageInterfac
 		return $autoloads;
 	}
 
+	/**
+	 * Take the autoloads array and return the psr4Map that contains the namespace path and version for each namespace.
+	 *
+	 * @param array      $autoloads Array of autoload settings defined defined by the packages.
+	 * @param bool       $scanPsrPackages Whether or not PSR packages should be converted into a classmap.
+	 * @param Filesystem $filesystem Filesystem class instance.
+	 * @param string     $vendorPath Path to the vendor directory.
+	 * @param string     $basePath Base Path.
+	 *
+	 * @return string $psr4Map
+	 */
+	private function getPsr4Map( array $autoloads, $scanPsrPackages, Filesystem $filesystem, $vendorPath, $basePath ) {
+		if ( $scanPsrPackages ) {
+			return 'array();' . PHP_EOL;
```

Comment:
That's a good point and would also solve your other comment about not checking the PSR-4 file for the autoloader first since it would't pass the `file_exists` check.

Reply:
Is there a reason to check the PSR-4 file for the autoloader first? Since the autoloader generator path should be in the classmap file in production, it seems like we should check the classmap first even if we don't generate the PSR-4 file.

Reply:
One of the things I was thinking about when writing this is that the `jetpack_autoload_psr4.php` file will be much smaller than the `jetpack_autoload_classmap.php` file. If we're saying the PSR-4 file is only generated when used then we can be confident that if it exists at all then the autoloader will be there and can avoid loading the classmap file unnecessarily.

Another thing I was thinking about when working on this is that we could probably do a bit of optimization around the manifest loading and the autoloader version reconciliation. We're currently loading the same manifest twice and I wonder if perhaps we should be storing the autoloader version using some different mechanism given the potentially large filesizes of manifest files for plugins with many dependencies.

Reply:
> We're currently loading the same manifest twice and I wonder if perhaps we should be storing the autoloader version using some different mechanism given the potentially large filesizes of manifest files for plugins with many dependencies.

One option would be to create a small file that returns an array containing just the autoloader version the the value of `$scanPsrPackages`. During the autoloader version reconciliation, we could determine the latest installed autoloader version and whether any plugin on the site is using a PSR-4 classmap. Were you thinking of something like that, or were you thinking about a different approach?

Reply:
That's exactly what I was thinking 😄 There do seem to be some other benefits too in that we can write information about the state of the autoloader at built-time and avoid some `file_exists` checks too.

Reply:
I've been thinking a bit about this and I think plugin load order might make it problematic. If a plugin with the `v2.0.2` autoloader initializes first then it won't know to check this file in the first place and will expect it in the classmap. I'm going to inject `Automattic\Jetpack\Autoloader\AutoloadGenerator` into the classmap to deal with this case and given that I think we can just rely on the lack of PSR-4 file presence as an indicator that the autoload is optimized.